ocaml/tests/test_pool_periodic_update_sync.ml

```diff
@@ -0,0 +1,284 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+open Test_highlevel
+open Pool_periodic_update_sync
+
+module TestUpdateSyncDelay = struct
+  let utc_offset_s = 0 (* UTC timezone for test *)
+
+  let utc8_offset_s = 8 * 60 * 60 (* UTC +8 timezone for test *)
+
+  let utc_1_offset_s = -1 * 60 * 60 (* UTC -1 timezone for test *)
+
+  let test_day_of_next_sync_1 tz_offset_s =
+    let now =
+      Ptime.of_date_time ((2023, 5, 4), ((13, 14, 15), tz_offset_s))
+      |> Option.get
+    in
+    let res = day_of_next_sync ~now ~tz_offset_s ~frequency:Daily in
+    let next_sync =
+      Ptime.of_date_time ((2023, 5, 5), ((0, 0, 0), tz_offset_s)) |> Option.get
+    in
+    Alcotest.(check bool)
+      "test_day_of_next_sync_1, daily" true
+      (Ptime.equal res next_sync)
```

Comment:
This testing looks much better, now we're testing the interesting cases!

Ptime can be tested natively instead of using the boolean checker:
```ocaml
  let ptime = Alcotest.testable Ptime.pp Ptime.equal
...
  let actual = day_of_next_sync ~now ~tz_offset_s ~frequency in
  Alcotest.check ptime "day_of_next" expected actual
```


Can the test cases be regrouped? There is a lot of repetition going on. We can have two test functions that actually check the properties, the data for timezones and test cases, and functions that combine the types of data and use the parametrized data to create all the tests. It can be a bit tricky because the name of the tests need to be constructed, but not too much:

for the next_day_sync

```ocaml
let timezone =
    [(0, "UTC"); (8 * 60 * 60, "UTC +8"); (-1 * 60 * 60, "UTC -1")]

  let next_day_test_cases =
    [
      {
        name= "daily"
      ; now= ((2023, 5, 4), (13, 14, 15))
      ; frequency= Daily
      ; expected= ((2023, 5, 5), (0, 0, 0))
      }
    ; {
        name= "daily, end of month"
      ; now= ((2023, 5, 31), (13, 14, 15))
      ; frequency= Daily
      ; expected= ((2023, 6, 1), (0, 0, 0))
      }
    ]

  let of_date_hour_tz d h tz_s = Ptime.of_date_time (d, (h, tz_s)) |> Option.get

  let test_next_day {name; now= d_a, h_a; frequency; expected= d_b, h_b}
      tz_offset_s () =
    let expected = of_date_hour_tz d_b h_b tz_offset_s in
    let now = of_date_hour_tz d_a h_a tz_offset_s in
    let actual = day_of_next_sync ~now ~tz_offset_s ~frequency in
    Alcotest.check ptime ("day_of_next_sync, " ^ name) expected actual

...

  let day_of_next_sync_test ({name; _} as test_case) =
    let test (tz, tz_name) =
      let name = Printf.sprintf "day_of_next_sync, %s, %s" name tz_name in
      (name, `Quick, test_next_day {test_case with name} tz)
    in
    List.map test timezones

  let test =
    List.concat_map day_of_next_sync_test next_day_test_cases
```

result:
```
  [OK]          pool_periodic_update_sync_periodic_update_sync          0   day_of_next_sync, daily, UTC.
  [OK]          pool_periodic_update_sync_periodic_update_sync          1   day_of_next_sync, daily, UTC +8.
  [OK]          pool_periodic_update_sync_periodic_update_sync          2   day_of_next_sync, daily, UTC -1.
  [OK]          pool_periodic_update_sync_periodic_update_sync          3   day_of_next_sync, daily, end of month, UTC.
  [OK]          pool_periodic_update_sync_periodic_update_sync          4   day_of_next_sync, daily, end of month, UTC +8.
  [OK]          pool_periodic_update_sync_periodic_update_sync          5   day_of_next_sync, daily, end of month, UTC -1.
```

Reply:
Thanks for the guide! Updated.